### ISSUE TITLE
WriteErrorException

### ISSUE THREAD

**Issue by ekt-** [TDabasinskas/Xiaomi#2]
First of all, absolutely great work.

Now the issue: Get-XiaomiGateway throws with this:

`+ CategoryInfo          : ProtocolError: (:) [Write-Error], WriteErrorException`
`+ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Get-XiaomiGateway`

Is it maybe related to the 16 symbol key you mention in the prerequisites, the one shown after activating developer mode? Where do I use that?

**Comment by tdabasinskas** [comment]
Hello @ekt-.

Do you you call [`Connect-XiaomiHome`](../blob/master/Functions/Connect-XiaomiHome.psm1) before calling [`Get-XiaomiGateway`](../blob/master/Functions/Get-XiaomiGateway.psm1)? Is there any error calling it or does it complete OK? If it's the latter, could you show the output (`Connect-XiaomiHome | FL`)?

As for the key, it is not currently being used,, as there are currently no methods that do write operations to the devices (the key is not required for the read-only operations). I'm waiting for [Xiaomi Socket](http://xiaomi-mi.com/smart-home/xiaomi-mi-smart-socket-plug/) to arrive and once I have it, I will try inplementing `Enable-XiaomiSocket` and `Disable-XiaomiSocket` commands, that would likely require the key.

Thanks for the feedback!

**Comment by ekt-** [comment]
yes, i'm calling connect | get. here the connect output

`Socket            : System.Net.Sockets.UdpClient`
`IsAlive           : True`
`LocalPort         : 9882`
`LocalEndpoint     : 0.0.0.0:9882`
`MulticastGroup    : 224.0.0.50`
`MulticastPeerPort : 4321`

**Comment by tdabasinskas** [comment]
Sorry for the delay - I've missed your reply.

Are you absolutely sure you followed all the steps described [on the third prerequisite in the README](../#prerequisites)? From what you reported (and knowing that the local UDP server is being created successfully), it seems that the gateway is unreachable. due to one of the following reasons:

1. You haven't properly enabled the developer mode.
2. Your router is filtering multicast traffic out.
3. There's a firewall blocking the access.

Once thing to note, that once the developer mode is enabled, you should be able to ping the gateway by IP. Try that.

**Comment by ekt-** [comment]
hei Tomas 
yes I believe i'm correctly in developer mode - and yes can I ping it
firewall is disabled - found a promising 'multicast' option in the router (Tomato), which was disabled, but even after enabling, same error

I plan to spend some time on it on the weekend. I'll keep you posted

**Comment by tdabasinskas** [comment]
Thanks, I'll be looking forward for the updates.

You can also edit [Get-XiaomiGateway.psm1 file](https://github.com/TDabasinskas/Xiaomi/blob/master/Functions/Get-XiaomiGateway.psm1) and replace lines 47-63 with the following code:

```powershell
$Connection.Send($endpoint, @{cmd = 'whois'});
$data = $Connection.Receive();
# Return the information about the gateway(s):
ForEach ($gateway in $data) {
    $gateways += [XiaomiGateway]::New($Connection, $data.sid, [IPAddress]::Parse($data.ip), $data.port);
}
Return $gateways;
```
This removed the `Try..Catch` block. so it would be easier for you to see why it fails.

By the way, is the firewall also off on your Windows machine?

**Comment by ekt-** [comment]
now that I think of it I've probably tested with windows firewall disabled and enabled multicast at the router level but at different times, never at the same time - so this could be the issue
will try tonight

**Comment by ekt-** [comment]
`Exception calling "Receive" with "1" argument(s): "A connection attempt failed`
`because the connected party did not properly respond after a period of time,`
`or established connection failed because connected host has failed to respond"`
`At C:\ProgramFiles\WindowsPowerShell\Modules\Xiaomi\0.1\Classes\XiaomiConnection.psm1:68 char:9`
`+         $bytes = $This.Socket.Receive([Ref]$This.LocalEndpoint);`

multicast on, windows firewall disabled. 

basically the receive fails here
`$bytes = $This.Socket.Receive([Ref]$This.LocalEndpoint);`

**Comment by ekt-** [comment]
naive question: could it be because the pc where the ps script runs is not connected via wifi?

**Comment by tdabasinskas** [comment]
Thanks for the data.

At least the exception indicated the problem is with getting the data to the PC, not in sending to to the network.

It worked for me via WiFi, so that should not be an issue. Just to be sure, both the gateway and the PC are connected to the same WiFi, in the same subnet?

Also, could you run `NETSTAT -ano -p UDP | Select-String "9882"` after running `Connect-XiaomiHome` and show the output?

**Comment by ekt-** [comment]
` UDP    0.0.0.0:9882           *:*                                    4472`

**Comment by tdabasinskas** [comment]
That seems right.

I'm actually running out of ideas now. I will let you know if I come up with something.

**Comment by ekt-** [comment]
About the wifi. The pc is NOT connected via wifi. Its on the same router but connected via cable. Subnet is the same tough

**Comment by tdabasinskas** [comment]
Oh, OK. I've misread it.

Could you try via WiFi just to rule that out?

**Comment by ekt-** [comment]
yep. will do. need to find a wifi dongle

**Comment by ekt-** [comment]
found the problem: multiple NICs
i had virtualbox and vmware installed. both tools install virtual nics. upd/multicast gets confused over which nic to listen.

(btw: i've discovered this by implementing a c# upd client for testing the multicast messages)

**Comment by tdabasinskas** [comment]
Hello @ekt- ,

I'm happy to hear you managed to get this working.

I guess the issue due to multiple NICs is related to the #54 line in [XiaomiConnection.psm1](../blob/master/Classes/XiaomiConnection.psm1#L54):
```powershell
$This.LocalEndpoint = [XiaomiConnection]::CreateEndpoint([IPAddress]::Any, $This.LocalPort);
```

I will try find a minute today to introduce an additional optional parameter to [Connect-XiaomiHome](../blob/master/Functions/Connect-XiaomiHome.psm1) called `-LocalIP`, so you can use it if you have multiple interfaces. Once specified, ` [XiaomiConnection]::CreateEndpoint` would use this IP address instead of `[IPAddress]::Any`. I think it should help.

**Comment by ekt-** [comment]
That would be useful. For the time being I've uninstalled virtualbox/vmware and everything now works nicely.

**Comment by TDabasinskas** [comment]
#585c958adb5425b7eaac2e0bc07f9e369085d465
